src/profile-logic/profile-data.js

```diff
@@ -150,6 +150,185 @@ export function getSampleCallNodes(
+
+    const thisNodeIndex = stackIndexToCallNodeIndex[stackIndex];
+    const thisFunc = callNodeTable.func[thisNodeIndex];
+
+    // For non-inverted trees, we compute the self time from the stacks' leaf nodes.
+    if (!isInvertedTree) {
+      if (thisNodeIndex === nodeIndex) {
+        pathTimings.selfTime.value += interval;
+
+        let implementation = 'native';
+        if (funcTable.isJS[thisFunc]) {
+          implementation = getJsImplementationForStack(stackIndex, thread);
+        }
+        accumulateDataToBreakdown(pathTimings.selfTime, implementation);
+      }
+
+      if (thisFunc === funcIndex) {
+        funcTimings.selfTime.value += interval;
+      }
+    }
+
+    // Use the callNodeTable to traverse the call node path and get various
+    // measurements.
+    // We don't use getCallNodePathFromIndex because we don't need the result
+    // itself, and it's costly to get.
+    let funcFound = false;
+    let pathFound = false;
+    let nextIndex;
+    for (
+      let currentIndex = thisNodeIndex;
+      currentIndex !== -1;
+      currentIndex = nextIndex
+    ) {
+      const currentFunc = callNodeTable.func[currentIndex];
+      nextIndex = callNodeTable.prefix[currentIndex];
+
+      // One of the parents is the exact passed path
+      if (currentIndex === nodeIndex) {
+        pathTimings.totalTime.value += interval;
+        pathFound = true;
+      }
+
+      // One of the parents' func is the same function as the passed path
+      // Note we could have the same function several times, so we need a
+      // boolean variable to prevent adding it more than once.
+      if (!funcFound && currentFunc === funcIndex) {
+        funcTimings.totalTime.value += interval;
+        funcFound = true;
+      }
+
+      // When the tree isn't inverted, we don't need to move further up the call
+      // node if we already found all the data.
+      if (!isInvertedTree && funcFound && pathFound) {
+        break;
+      }
+
+      // But for inverted trees, the selfTime needs to be counted on the root
+      // node. This is the root node if nextIndex is -1.
+      if (isInvertedTree && nextIndex === -1) {
+        // If the root node is the passed path
+        if (currentIndex === nodeIndex) {
+          pathTimings.selfTime.value += interval;
+        }
+
+        // If the root node is the same function as the passed path
+        if (currentFunc === funcIndex) {
+          funcTimings.selfTime.value += interval;
+        }
+
+        // Contributing the implementation information if the passed path was
+        // found in this stack _and_ if the root node (the one actually holding
+        // the self time) is a JS frame.
+        // This allows to see how the implementation timings is spread among
+        // callers.
+        if (pathFound) {
+          let implementation = 'native';
+          if (funcTable.isJS[currentFunc]) {
+            implementation = getJsImplementationForStack(stackIndex, thread);
```

Comment:
I gave a lot of thought to this since yesterday, and I'm not so sure I should use the stackIndex here; rather I believe I should use the stackIndex for the root node... but I don't have it here because I used the call node table.
So I think I have to rewrite the algorithm using the stack table instead of the call node table just so that I can access the frame information for the root node.